formats/json/jvmMain/src/kotlinx/serialization/json/internal/CharArrayPool.kt

```diff
@@ -13,20 +17,37 @@ internal object CharArrayPool {
         System.getProperty("kotlinx.serialization.json.pool.size").toIntOrNull()
     }.getOrNull() ?: 1024 * 1024 // 2 MB seems to be a reasonable constraint, (1M of chars)
 
-    fun take(): CharArray {
+    protected fun take(size: Int): CharArray {
```

Comment:
I think it's better to take the size in the class constructor rather than in the function call.

In fact, this is the case for each instance of the pool right now.
Another reason is a potential error if I call like this
```
val a = take(100)
release(a)
val b = take(100)
release(b)
val c = take(200)
c[101] // !!!
```
The API does not advise in any way that you should not do this.

If you add a size to the constructor, then you can also add a check for the size of the array inside the `releaseImpl`

Reply:
This is a good remark for any general-purpose API, but in this situation, we are dealing with well-tailored internal pools for one specific purpose. Also, it will force us to introduce a few more bytecodes -- for class with ctor, for outer class that will store the instance etc.

Reply:
In the byte code, only a new field with the int type will be added (the parameter will be transferred from the `take` method to the `init` method), and in the child class, only the value from the parameter will be setted to the child constructor.

I think this is a good tradeoff so that if someone else refactor the code after a while, they don't add an unobvious bug.

Reply:
Given that this is a protected method, I don't see that it is necessary for now, but the idea overall is a valid concern.

Reply:
The method is indeed protected and sued as a **base** for the implementation.
None of the implementations are prone to the potential bug described above and I prefer to avoid overcomplications in a 50 lines of fully internal code that has a single purpose